fdbserver/DataDistribution.actor.cpp

```diff
@@ -1512,219 +1631,690 @@ ACTOR Future<Void> ddGetMetrics(GetDataDistributorMetricsRequest req,
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		ASSERT(audit != nullptr);
+		audit->actors.add(loadAndDispatchAudit(self, audit, audit->coreState.range));
+		TraceEvent(SevInfo, "DDAuditStorageCoreScheduled", self->ddId)
+		    .detail("Context", context)
+		    .detail("AuditID", audit->coreState.id)
+		    .detail("Range", audit->coreState.range)
+		    .detail("AuditType", audit->coreState.getType())
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		wait(audit->actors.getResult()); // goto exception handler if any actor is failed
+		if (audit->foundError) {
+			audit->coreState.setPhase(AuditPhase::Error);
+		} else if (audit->anyChildAuditFailed) {
+			audit->anyChildAuditFailed = false;
+			throw audit_storage_failed(); // retry
+		} else {
+			audit->coreState.setPhase(AuditPhase::Complete);
+		}
+		TraceEvent(SevVerbose, "DDAuditStorageCoreGotResult", self->ddId)
+		    .detail("Context", context)
+		    .detail("AuditState", audit->coreState.toString())
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		wait(persistAuditState(self->txnProcessor->context(),
+		                       audit->coreState,
+		                       "AuditStorageCore",
+		                       lockInfo,
+		                       self->context->isDDEnabled()));
+		TraceEvent(SevVerbose, "DDAuditStorageCoreSetResult", self->ddId)
+		    .detail("Context", context)
+		    .detail("AuditState", audit->coreState.toString())
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		removeAuditFromAuditMap(self,
+		                        audit->coreState.getType(),
+		                        audit->coreState.id,
+		                        audit->coreState.getPhase(),
+		                        audit->retryCount); // remove audit
+
+		TraceEvent(SevInfo, "DDAuditStorageCoreEnd", self->ddId)
+		    .detail("Context", context)
+		    .detail("AuditID", auditID)
+		    .detail("AuditType", auditType)
+		    .detail("Range", audit->coreState.range)
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+	} catch (Error& e) {
+		TraceEvent(SevDebug, "DDAuditStorageCoreError", self->ddId)
+		    .errorUnsuppressed(e)
+		    .detail("Context", context)
+		    .detail("AuditID", auditID)
+		    .detail("RetryCount", currentRetryCount)
+		    .detail("AuditType", auditType)
+		    .detail("Range", audit->coreState.range)
+		    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+		if (e.code() == error_code_actor_cancelled || e.code() == error_code_movekeys_conflict) {
+			throw e;
+		} else if (audit->retryCount < SERVER_KNOBS->AUDIT_RETRY_COUNT_MAX && e.code() != error_code_not_implemented) {
+			audit->retryCount++;
+			audit->actors.clear(true);
```

Comment:
can we use audit->cancel() here?

Reply:
No. audit->cancel() aims to cancel the entire audit, which includes the audit leader (auditStorageCore) and workers (audit->actors, e.g., schedulers and doAudits). Here, we only want to cleanup the existing workers.